### PR TITLE
feat: add support for generating uuids in SQL

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Closes https://github.com/sequelize/sequelize/issues/13224

This PR adds `sql.uuidV1` and `sql.uuidV4`, which are aimed to be replacements for `DataTypes.UUIDV1` and `DataTypes.UUIDV4` respectively.

Unlike the DataType version, these variants set the default value in the database schema when possible, but fall back to generating the default value in JavaScript. This makes them useful in both models and migrations (assuming the dialect provides a way to generate UUIDs):

```ts
User.init({
  attr: {
    type: DataTypes.UUID,
    allowNull: false,
    defaultValue: sql.uuidV4,
  },
}, { sequelize });
```

They are also usable in any raw SQL clause, if the dialect supports generating UUIDs. The following:

```ts
sequelize.query(sql`INSERT INTO table (id) VALUES (${sql.uuidV4});`);
```

Would generate the following query, in postgres:

```sql
INSERT INTO table (id) VALUES (gen_random_uuid());
```

Sequelizes tries to determine if the dialect supports generating UUIDs, but can be wrong (generating UUID v1 in postgres requires the `uuid-ossp` extension, which we assume is installed).

If Sequelize gets it wrong, the user can fall back to the JavaScript version themselves by using `sql.uuidV1.asJavaScript` or `sql.uuidV4.asJavaScript`:

```ts
User.init({
  attr: {
    type: DataTypes.UUID,
    allowNull: false,
    defaultValue: sql.uuidV4.asJavaScript,
  },
}, { sequelize });
```

This is done by expanding `DialectAwareFn` to make it capable of generating JavaScript values.

---

`DataTypes.UUIDV4` and `DataTypes.UUIDV1` were also used to alter the validation of UUIDs, restricting them to either V4 or V1 instead of any UUID version.

`sql.uuidVx` does not do that, as the concern is separate and you should be able to restrict the UUID version if you don't set a default value, or if you used a different default value.

Instead, `DataTypes.UUID` now has 2 extensions: `DataTypes.UUID.V1` and `DataTypes.UUID.V4`, which control the validation of the UUID version.

---

In future PRs, we could look into migrating `DataTypes.NOW` to do the same, but it may be wise to wait for Temporal, so we're not stuck with legacy behavior like with `DataTypes.NOW`. We can also look into migrating `sequelize.random()` into `sql.random`

### CODE REVIEW

**Comment by ephys** [comment]
Drafting because I forgot to add value format validation to the new system

File: packages/core/test/integration/data-types/data-types.test.ts [hunk 1]
```diff
@@ -1229,6 +1229,90 @@ describe('DataTypes', () => {
     });
   });
 
+  describe('UUID default values', () => {
+    beforeAll2(async () => {
+      if (dialect.name === 'postgres') {
+        await sequelize.query('CREATE EXTENSION IF NOT EXISTS "uuid-ossp"');
```
Comment (WikiRik) [inline packages/core/test/integration/data-types/data-types.test.ts:1235]: Is this packaged by default in the Docker image we use?

File: packages/core/test/unit/query-interface/create-table.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import { createSequelizeInstance, expectsql, sequelize } from '../../support';
+
+const dialect = sequelize.dialect;
+
+describe('QueryInterface#createTable', () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  it('supports sql.uuidV4 default values', async () => {
```
Comment (WikiRik) [inline packages/core/test/unit/query-interface/create-table.test.ts:13]: For this and the following test; do we have a policy on testing on earlier (supported) versions?

**Review by WikiRik — COMMENTED** [review]
Nice stuff!

**Comment by WikiRik** [comment]
> In future PRs, we could look into migrating DataTypes.NOW to do the same, but it may be wise to wait for Temporal, so we're not stuck with legacy behavior like with DataTypes.NOW. 

Agree with waiting until Temporal lands

File: packages/core/test/integration/data-types/data-types.test.ts (hunk 1 above)
Comment (ephys) [reply packages/core/test/integration/data-types/data-types.test.ts:1235]: Seems to be. I haven't encountered a postgres distribution that didn't include it

File: packages/core/test/unit/query-interface/create-table.test.ts (hunk 2 above)
Comment (ephys) [reply packages/core/test/unit/query-interface/create-table.test.ts:13]: I think we should cover all possible branches, done

File: packages/core/src/utils/deprecations.ts [hunk 3]
```diff
@@ -25,3 +25,4 @@ export const noSqlJson = deprecate(noop, 'The json() function used to generate J
 export const alwaysQuoteIdentifiers = deprecate(noop, 'Setting "quoteIdentifiers" to false is unsafe and it will be removed in v8.', 'SEQUELIZE0023');
 export const showAllToListSchemas = deprecate(noop, 'Do not use "showAllSchemas". Use QueryInterface#listSchemas instead.', 'SEQUELIZE0024');
 export const showAllToListTables = deprecate(noop, 'Do not use "showAllTables". Use QueryInterface#listTables instead.', 'SEQUELIZE0025');
+export const noDataTypesUuid = deprecate(noop, 'Do not use DataTypes.UUIDV1 or DataTypes.UUIDV2. Use sql.uuidV1 or sql.uuidV4 instead.', 'SEQUELIZE0026');
```
[the remark this answers is not in the document]
Comment (ephys) [reply packages/core/src/utils/deprecations.ts]: Done

**Review by WikiRik — APPROVED** [review]

**Pull request merged**